# Retrieval Layer — Stage 1: Chunking the Corpus

Builds the retrieval half of a RAG system over the SDE methodology decision record.

**Why this exists.** The existing valuation pipeline formats the entire practice dataset
into one string and sends all of it on every question. That is correct for numbers — you
want the *exact* revenue figure, not the four rows most similar to the phrase "revenue."
But it left a gap: when the model explained *why* an expense qualified as an add-back, it
had nothing to draw on. Those rules lived in regex patterns and hardcoded constants, where
the model could not see them.

This notebook makes the rules retrievable. The pipeline stays split:

| Question | Path | Why |
|---|---|---|
| "What was 2023 revenue?" | existing deterministic lookup | exactness matters; retrieval would be a downgrade |
| "Why is automobile only 25%?" | retrieval (this notebook) | the answer is a passage, not a cell |

Kept separate from the main pipeline because retrieval is independently testable — it
needs no API key, no CSVs, no practice data. Only the markdown file.

## 1. Chunking

Retrieval searches **chunks**, not documents. Each chunk becomes one vector, and a chunk
covering four topics produces a vector that is a blurred average of all four — close to
nothing in particular, so it is retrieved for none of them. Too small fails the other way:
"$9,114.00 in FY2022" is precise and meaningless without the sentence defining it.

The target is a passage that answers one question completely.

**Strategies and tradeoffs:**

| Approach | Use when |
|---|---|
| Fixed size (every N chars) | text has no structure; accepts mid-sentence cuts |
| Paragraph | structure exists but headings don't |
| **Section / heading** | the document was written with meaningful headings |
| Semantic (detect topic shift by embedding similarity) | structure is unreliable and accuracy justifies the cost |

Section-based here, because the decision record was drafted so each section answers one
question. That structure *is* the chunking; the code only has to read it.

**Breadcrumbs.** `## 5. Known methodology issues` is a bare heading — its body is entirely
subsections. Chunked naively, its children survive but lose their parent, producing a
chunk labeled only "FY2022" with nothing in it saying *owner compensation*.

So each chunk carries its parent heading, written **into the chunk text** rather than
stored beside it. Metadata kept in a separate field is not embedded and cannot affect
whether a chunk is found. This generalizes: anything that should influence retrieval must
be inside the embedded text.

In [2]:
## Cell 1 — parse the document into chunks

import re
from pathlib import Path

DOC = Path("SDE_Method_Decision_Record.md")

# encoding="utf-8" is not optional. Without it Python falls back to the platform
# default (cp1252 on Windows) and every em dash arrives as "—" — which then gets
# embedded, degrading the vectors silently.
text = DOC.read_text(encoding="utf-8")

# Split BEFORE each level-2 or level-3 heading.
#   \n(?=...)  lookahead: consume the newline, leave the "##" attached to its section
#   #{2,3}     "##" or "###" only — the "#" title line is not a section
chunks, h2 = [], None

for part in re.split(r'\n(?=#{2,3} )', text):
    part = part.strip()
    if not part.startswith('#'):
        continue                                  # drops the title block

    line    = part.split('\n', 1)[0]
    level   = len(line) - len(line.lstrip('#'))
    heading = line.lstrip('# ').strip()

    if level == 2:
        h2, breadcrumb = heading, heading
    else:
        breadcrumb = f"{h2} > {heading}" if h2 else heading

    body = part.split('\n', 1)[1].strip() if '\n' in part else ''
    if not body:
        continue                                  # bare heading: survives in its children's breadcrumbs

    chunks.append({"id": f"sde-{len(chunks):02d}",
                   "heading": breadcrumb,
                   "text": f"{breadcrumb}\n\n{body}"})

print(f"{len(chunks)} chunks\n")
for c in chunks:
    flag = "  <-- oversized" if len(c["text"]) > 2000 else ""
    print(f"  {c['id']}  {len(c['text']):>5}  {c['heading'][:65]}{flag}")

15 chunks

  sde-00    353  SDE Methodology — Decision Record
  sde-01   1179  1. Engagement purpose
  sde-02    743  2. Source documents and their roles
  sde-03    455  3. The formula
  sde-04    502  3. The formula > Verification — reconciles to the cent, all four 
  sde-05   1348  4. Add-back categories
  sde-06    357  5. Known methodology issues > 5.1 Automobile — 25% personal use i
  sde-07   2486  5. Known methodology issues > 5.2 Owner compensation — mismatched  <-- oversized
  sde-08    404  5. Known methodology issues > 5.3 Multi-owner benchmark
  sde-09    288  5. Known methodology issues > 5.4 Partial year
  sde-10   1125  6. Corrected figures
  sde-11    490  7. Owner compensation — full reconciliation > FY2022
  sde-12   1147  7. Owner compensation — full reconciliation > FY2023 — formula co
  sde-13   1302  8. Open items
  sde-14    523  9. What this document is for


## 2. Splitting oversized sections

The owner-compensation section runs past 2,000 characters because it contains both the
reasoning (why employer payroll tax does not belong in the numerator) and the resulting
figures. Those answer different questions, so fusing them means every query retrieves both
and half of what reaches the model is irrelevant.

Splitting at a fixed character count would cut mid-table, so segments break only at
paragraph boundaries.

**Two thresholds, not one.** 2,000 triggers a split; 1,400 sizes the results. If both were
2,000, a 2,100-character section would yield one full segment and a useless 100-character
runt.

Sizes are in characters. Embedding models cap on *tokens* and silently truncate past the
limit — worse than an error, since nothing reports it. At roughly 4 characters per token
for English, 1,400 characters is about 350 tokens, well under any model's ceiling, so
retrieval quality binds before truncation does.

In [3]:
## Cell 2 — split oversized chunks at paragraph boundaries

MAX_SEGMENT = 1400     # target size for a split segment
OVERSIZED   = 2000     # threshold above which a chunk is split at all

def pack(paragraphs, limit):
    """Group whole paragraphs into segments under `limit` characters.

    A single paragraph longer than `limit` becomes its own oversized segment rather
    than being cut — a slightly-too-long coherent passage beats a severed one.
    """
    segments, cur = [], []
    for p in paragraphs:
        candidate = cur + [p]
        # + 2*len(cur) accounts for the "\n\n" rejoining each pair
        if cur and sum(len(x) for x in candidate) + 2 * len(cur) > limit:
            segments.append(cur)
            cur = [p]
        else:
            cur = candidate
    if cur:
        segments.append(cur)
    return segments


final = []
for c in chunks:
    if len(c["text"]) <= OVERSIZED:
        final.append(c)
        continue

    body = c["text"].split("\n\n", 1)[1]          # drop breadcrumb before splitting
    for i, seg in enumerate(pack(body.split("\n\n"), MAX_SEGMENT)):
        label = c["heading"] if i == 0 else f"{c['heading']} (cont. {i+1})"
        final.append({"id": None, "heading": label,
                      "text": f"{label}\n\n" + "\n\n".join(seg)})

for i, c in enumerate(final):                      # ids must stay contiguous after splitting
    c["id"] = f"sde-{i:02d}"

chunks = final
print(f"{len(chunks)} chunks, max {max(len(c['text']) for c in chunks)} chars")

16 chunks, max 1390 chars


## 3. Verify and save

Sizes cannot tell you whether a chunk is about one thing — that requires reading the
boundaries. The first segment of a split section should hold the rule; the continuation
should hold the figures. A cut landing mid-table is invisible until retrieval quietly
returns the wrong half.

Saving to JSON freezes the chunk set. If chunking drifts between runs, retrieval quality
drifts with it and you cannot tell whether a change came from a new embedding model or
from different chunks.

In [4]:
## Cell 3 — inspect boundaries, check invariants, persist

import json

for c in chunks:
    if "(cont." in c["heading"] or "5.2" in c["heading"]:
        print(f"--- {c['id']} · {c['heading'][:70]}")
        preview = c["text"].split("\n\n", 1)[1][:260]
        print("    " + preview.rsplit(" ", 1)[0].replace("\n", "\n    ") + " ...\n")

assert all(len(c["text"]) <= 2000 for c in chunks),      "a chunk is still oversized"
assert len({c["id"] for c in chunks}) == len(chunks),    "duplicate chunk ids"
assert all(c["heading"] in c["text"] for c in chunks),   "breadcrumb missing from text"

OUT = Path("sde_chunks.json")
OUT.write_text(json.dumps(chunks, indent=2, ensure_ascii=False), encoding="utf-8")
print(f"OK — {len(chunks)} chunks, {sum(len(c['text']) for c in chunks):,} chars -> {OUT}")

--- sde-07 · 5. Known methodology issues > 5.2 Owner compensation — mismatched basi
    The numerator was **loaded** cost (wages + employer Social Security). The benchmark,
    $195,700, is a **gross salary** figure. Comparing cost against salary overstates excess by
    the employer tax — which is not excess at all, since employing *any* dentist at ...

--- sde-08 · 5. Known methodology issues > 5.2 Owner compensation — mismatched basi
    The correction removes exactly the capped employer Social Security in each year
    ($9,114.00 in FY2022, $9,932.40 in FY2023). FY2024 and FY2025 are unaffected because owner
    salary had already fallen below the benchmark.
    
    **Conclusion unchanged.** After the ...

OK — 16 chunks, 12,794 chars -> sde_chunks.json


---

# Stage 2 — Embedding

Retrieval compares *meaning*, not words. A question asking why payroll tax was excluded
should find a chunk discussing employer Social Security even though the two share almost
no vocabulary. That requires converting text into vectors positioned so that nearby means
similar in meaning.

**Model: `all-MiniLM-L6-v2`.** ~90 MB, Apache-2.0, CPU-fast, 384 dimensions.
`all-mpnet-base-v2` scores higher on benchmarks at 768 dimensions and ~5x the size, but on
a 16-chunk corpus the difference is undetectable — retrieval failures at this scale come
from chunking, not from the model. `bge-small-en-v1.5` benchmarks above MiniLM but requires
prefixing every *query* with an instruction string, and omitting it degrades results
silently. Not a good first build.

**Driving the model directly rather than calling `.encode()`.** `sentence-transformers`
offers a one-line shortcut that does everything below. It is the right choice in
production and teaches nothing, so the pipeline is written out: tokenize, forward pass,
pool, normalize. Every step is a place things go wrong.

## 4. Setup

```
pip install torch transformers
```

Weights download automatically on first use (90 MB, cached in `~/.cache/huggingface`).
Every run after the first is offline. No account, no key, no per-call cost.

In [5]:
## Cell 4 — load the tokenizer and model

import json, torch
from pathlib import Path
from transformers import AutoTokenizer, AutoModel

MODEL_NAME = "sentence-transformers/all-MiniLM-L6-v2"

chunks = json.loads(Path("sde_chunks.json").read_text(encoding="utf-8"))

# A tokenizer and its model are a matched pair. Each model was trained against one
# specific vocabulary and one set of special tokens; pairing a model with a different
# tokenizer produces plausible-looking garbage rather than an error.
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
model     = AutoModel.from_pretrained(MODEL_NAME)

# eval() disables dropout and other training-only behaviour. Without it the same input
# yields slightly different vectors on each call — a genuinely confusing bug, since
# nothing fails, results are just quietly non-reproducible.
model.eval()

device = "cuda" if torch.cuda.is_available() else "cpu"
model.to(device)

print(f"{MODEL_NAME}")
print(f"device            : {device}")
print(f"hidden size       : {model.config.hidden_size}")
print(f"max input tokens  : {tokenizer.model_max_length}")
print(f"chunks to embed   : {len(chunks)}")

sentence-transformers/all-MiniLM-L6-v2
device            : cpu
hidden size       : 384
max input tokens  : 512
chunks to embed   : 16


## 5. Tokenizing

Models consume integer token ids, not text. The tokenizer splits each string into subword
units and maps them to ids.

Every sequence in a batch must be the same length to form a rectangular tensor, so shorter
ones get **padded**. Padding is meaningless filler, and the `attention_mask` records which
positions are real (1) and which are padding (0). That mask matters more than it looks —
see the next cell.

`truncation=True` caps input at the model's limit (512 tokens here). Without it, longer
input raises an error in some versions and is silently cut in others. Our chunks top out
around 350 tokens, so nothing should truncate — but the flag makes that a guarantee rather
than an assumption.

In [6]:
## Cell 5 — tokenize the corpus

texts = [c["text"] for c in chunks]

encoded = tokenizer(
    texts,
    padding=True,        # pad every sequence to the longest in the batch
    truncation=True,     # cap at the model's maximum input length
    max_length=512,
    return_tensors="pt", # return PyTorch tensors rather than lists
)
encoded = {k: v.to(device) for k, v in encoded.items()}

ids, mask = encoded["input_ids"], encoded["attention_mask"]
print(f"input_ids      : {tuple(ids.shape)}   (chunks x padded_length)")
print(f"attention_mask : {tuple(mask.shape)}")
print(f"real tokens per chunk: min {int(mask.sum(1).min())}, max {int(mask.sum(1).max())}")

# Confirm nothing was truncated: if a chunk hit the cap exactly, content was lost.
if int(mask.sum(1).max()) >= 512:
    print("WARNING: at least one chunk hit the token limit and was truncated")

input_ids      : (16, 450)   (chunks x padded_length)
attention_mask : (16, 450)
real tokens per chunk: min 74, max 450


## 6. Forward pass and pooling

The model returns one vector **per token** — shape `(chunks, tokens, 384)`. We need one
vector per chunk, so the token vectors get collapsed into a single vector. That is pooling.

**Mean pooling with the mask is the part to get right.** Averaging naively across the
token axis includes the padding positions and drags every short chunk's vector toward
whatever the pad token happens to encode. Nothing errors; results are just quietly wrong,
and worse for shorter chunks. The mask is expanded to the hidden dimension, zeroes out
padding, and the sum is divided by the count of *real* tokens only.

**`torch.no_grad()`** turns off gradient tracking. PyTorch normally records every
operation so gradients can be computed for training — pure overhead during inference, in
both memory and speed. This is inference, so it is switched off.

**Normalization.** Scaling each vector to unit length makes cosine similarity equal to a
plain dot product, which is why the search step later is a single matrix multiply.

In [7]:
## Cell 6 — forward pass, mean pooling, normalization

def mean_pool(hidden_states, attention_mask):
    """Collapse per-token vectors into one vector per sequence, ignoring padding.

    hidden_states : (batch, tokens, hidden)
    attention_mask: (batch, tokens)  1 = real token, 0 = padding
    returns       : (batch, hidden)
    """
    # (batch, tokens) -> (batch, tokens, hidden) so it lines up with hidden_states
    m = attention_mask.unsqueeze(-1).expand(hidden_states.size()).float()
    summed = (hidden_states * m).sum(dim=1)          # zero out padding, then sum
    counts = m.sum(dim=1).clamp(min=1e-9)            # divide by REAL tokens only
    return summed / counts


with torch.no_grad():                                 # inference: no gradients needed
    output = model(**encoded)

hidden = output.last_hidden_state
print(f"last_hidden_state : {tuple(hidden.shape)}   (chunks x tokens x hidden)")

pooled = mean_pool(hidden, mask)
print(f"after pooling     : {tuple(pooled.shape)}")

embeddings = torch.nn.functional.normalize(pooled, p=2, dim=1)
print(f"after normalizing : {tuple(embeddings.shape)}")
print(f"vector norms      : all 1.0 -> {bool(torch.allclose(embeddings.norm(dim=1), torch.ones(len(chunks), device=device)))}")

# Demonstrates why the mask is not optional: naive pooling gives different vectors.
naive = hidden.mean(dim=1)
naive = torch.nn.functional.normalize(naive, p=2, dim=1)
drift = (embeddings - naive).norm(dim=1)
print(f"\nmasked vs naive pooling, mean drift: {float(drift.mean()):.4f}  (0.0 would mean the mask changed nothing)")

last_hidden_state : (16, 450, 384)   (chunks x tokens x hidden)
after pooling     : (16, 384)
after normalizing : (16, 384)
vector norms      : all 1.0 -> True

masked vs naive pooling, mean drift: 0.7474  (0.0 would mean the mask changed nothing)


## 7. Sanity-checking the vectors

Embeddings can be the wrong shape and still look fine, so check them against something you
already know: `sde-07` and `sde-08` are two halves of the same section and should be among
each other's nearest neighbours. If they are not, something upstream is broken.

With unit-length vectors, `E @ E.T` produces the full similarity matrix in one matrix
multiply. Values run from -1 to 1; on a single-topic corpus everything tends to sit high,
so what matters is *relative* ranking, not absolute values.

In [8]:
## Cell 7 — inspect nearest neighbours

sim = (embeddings @ embeddings.T).cpu()
sim.fill_diagonal_(-1)                # exclude self-matches

for i in [c["id"] for c in chunks].index("sde-07"), [c["id"] for c in chunks].index("sde-08"):
    best = sim[i].argsort(descending=True)[:3]
    print(f"{chunks[i]['id']}  {chunks[i]['heading'][:58]}")
    for j in best:
        print(f"     {sim[i][j]:.3f}  {chunks[j]['id']}  {chunks[j]['heading'][:52]}")
    print()

sde-07  5. Known methodology issues > 5.2 Owner compensation — mis
     0.729  sde-11  6. Corrected figures
     0.682  sde-13  7. Owner compensation — full reconciliation > FY2023
     0.679  sde-08  5. Known methodology issues > 5.2 Owner compensation

sde-08  5. Known methodology issues > 5.2 Owner compensation — mis
     0.762  sde-13  7. Owner compensation — full reconciliation > FY2023
     0.718  sde-12  7. Owner compensation — full reconciliation > FY2022
     0.688  sde-11  6. Corrected figures



## 8. Persisting the vectors

Saved as `.npy` rather than JSON: floats survive exactly, and the file is a fraction of the
size.

**Store the model name alongside them.** Vectors from different models are not comparable —
embedding a query with one model and searching an index built by another returns nonsense
with no error. Recording which model produced the index makes that mismatch detectable
instead of mysterious.

In [9]:
## Cell 8 — save vectors and index metadata

import numpy as np

vectors = embeddings.cpu().numpy().astype("float32")
np.save("sde_vectors.npy", vectors)

Path("sde_index_meta.json").write_text(json.dumps({
    "model": MODEL_NAME,
    "dim": int(vectors.shape[1]),
    "count": int(vectors.shape[0]),
    "chunk_ids": [c["id"] for c in chunks],
}, indent=2), encoding="utf-8")

print(f"saved {vectors.shape[0]} x {vectors.shape[1]} vectors ({vectors.nbytes/1024:.1f} KB)")
print(f"index model: {MODEL_NAME}")

saved 16 x 384 vectors (24.0 KB)
index model: sentence-transformers/all-MiniLM-L6-v2


## Next: Stage 3 — retrieval

Embed an incoming question with the *same* model, compare it against all 16 vectors with
one dot product, and return the top matches.

**No vector database.** FAISS, Chroma, and Pinecone make *approximate* search fast once
exhaustive comparison becomes impractical — somewhere around a hundred thousand vectors.
At 16, a full comparison is a 16x384 matrix multiply measured in microseconds. Adding one
here would buy a dependency, a persistence format, and an abstraction over `numpy.dot`, in
exchange for nothing.

In [10]:
## Cell 9 — search: embed a question, rank chunks by similarity

def embed(texts):
    """Tokenize -> forward pass -> mask-aware mean pool -> normalize.
    Identical to cells 5-6. Queries MUST go through the same model as the chunks."""
    enc = tokenizer(texts, padding=True, truncation=True, max_length=512, return_tensors="pt")
    enc = {k: v.to(device) for k, v in enc.items()}
    with torch.no_grad():
        out = model(**enc)
    pooled = mean_pool(out.last_hidden_state, enc["attention_mask"])
    return torch.nn.functional.normalize(pooled, p=2, dim=1)


def search(question, k=3):
    q = embed([question])              # (1, 384)
    scores = (q @ embeddings.T)[0]     # compare against all 16 -> (16,)
    top = scores.argsort(descending=True)[:k]
    return [(float(scores[i]), chunks[i]) for i in top]


for score, c in search("Why is the automobile add-back only 25 percent?"):
    print(f"{score:.3f}  {c['id']}  {c['heading'][:60]}")

0.546  sde-06  5. Known methodology issues > 5.1 Automobile — 25% personal 
0.281  sde-10  5. Known methodology issues > 5.4 Partial year
0.275  sde-04  3. The formula > Verification — reconciles to the cent, all 


## 9. Dual-index search

**The problem.** Searching `"open items"` did not find the chunk titled "8. Open items" —
even though those exact words are in it.

**Why.** Each chunk becomes one vector by averaging all its words together. That chunk is
about 350 words long, and only 3 of them are the heading. So the heading is less than 1% of
the average. The other 99% is dollar amounts and surnames, and that is what the chunk ends
up "about."

Embedding just the heading by itself scored 0.850 instead of 0.098. Same model, same words —
the difference is that nothing drowned it out.

This also rules out the obvious fix. Adding a summary sentence to each chunk would just be
one more line getting averaged away with everything else.

**The fix: give each chunk two vectors instead of one.**

| Index | Made from | Good at finding |
|---|---|---|
| body | the whole chunk | details — `"Hassenplug compensation"`, `"$9,114"` |
| heading | just the title | topics — `"open items"`, `"partial year"` |

Each chunk gets scored both ways, and we keep the higher score. They cover each other's
gaps: `"Hassenplug"` is in no title, so only the body finds it. `"open items"` is drowned
in the body, so only the heading finds it.

**One catch:** headings always score higher than bodies, because short text matches short
text. A 0.850 heading match is not "better" than a 0.546 body match — they are measured
differently. So never write a rule like "ignore anything below 0.4." Trust which chunk came
first, not the number itself.

**Known limitation.** `"What is still unresolved?"` still fails. The model does not know
that *unresolved* and *open items* mean the same thing. That is a weakness of small
embedding models, and no amount of reorganizing fixes it.

In [39]:
## Cell 10 — dual-index search with client isolation

head_vecs = embed([c["heading"] for c in chunks])

# Tag each chunk with the client it belongs to. None = shared method corpus,
# retrievable for every client. Sections 2, 3 and 5 are general method; the rest
# are Roselle-specific figures.
for c in chunks:
    c["client"] = None if c["heading"].startswith(("2.", "3.", "5.")) else "roselle"


def search(question, k=3, client="roselle"):
    q    = embed([question])
    body = (q @ embeddings.T)[0]
    head = (q @ head_vecs.T)[0]
    best = torch.maximum(body, head)

    # Chunks belonging to another client are pushed to -inf so they can never rank.
    allowed = [i for i, c in enumerate(chunks) if c["client"] in (None, client)]
    gate = torch.full_like(best, float("-inf"))
    gate[allowed] = 0.0
    best = best + gate

    top = best.argsort(descending=True)[:k]
    return [(float(best[i]), "head" if head[i] > body[i] else "body", chunks[i]) for i in top]


# Confirm the tagging landed where you expect
for c in chunks:
    print(f"  {c['id']}  {str(c['client']):8}  {c['heading'][:58]}")

  sde-00  roselle   SDE Methodology — Decision Record
  sde-01  roselle   1. Engagement purpose
  sde-02  None      2. Source documents and their roles
  sde-03  None      3. The formula
  sde-04  None      3. The formula > Verification — reconciles to the cent, al
  sde-05  roselle   4. Add-back categories
  sde-06  None      5. Known methodology issues > 5.1 Automobile — 25% persona
  sde-07  None      5. Known methodology issues > 5.2 Owner compensation — mis
  sde-08  None      5. Known methodology issues > 5.2 Owner compensation — mis
  sde-09  None      5. Known methodology issues > 5.3 Multi-owner benchmark
  sde-10  None      5. Known methodology issues > 5.4 Partial year
  sde-11  roselle   6. Corrected figures
  sde-12  roselle   7. Owner compensation — full reconciliation > FY2022
  sde-13  roselle   7. Owner compensation — full reconciliation > FY2023 — for
  sde-14  roselle   8. Open items
  sde-15  roselle   9. What this document is for


## 10. Generation — answering from the retrieved chunks

Retrieval finds the right chunks. It does not answer the question. This is the second half
of RAG: hand those chunks to a language model and have it write an answer *from them*.

The model runs on someone else's servers, reached over the internet. So unlike everything
above, this part can fail for reasons that have nothing to do with your code — the service
can be busy, the key can be wrong, the network can drop. That is why it is split into three
cells instead of one.

| Cell | Does | Run it |
|---|---|---|
| 11A | stores your key and picks the model | once per session |
| 11B | sends 5 tokens to check the connection | whenever something breaks |
| 11C | the actual question-answering | as often as you like |

Keeping the key in its own cell means re-running the question does not re-prompt you for it.
Keeping the connection test separate means that when something fails, you can tell whether
the problem is the service or your code — without that, a network hiccup looks identical to
a bug you wrote.

### 11A — configuration

The API key identifies your account. It is read from an environment variable if one exists,
and otherwise typed in at a prompt. Either way it is never written into the notebook, so
the file can be shared or committed without leaking a credential. Pasting a key directly
into a cell is the single most common way people accidentally publish one.

`MODEL` is the identifier of the model you are calling. Changing this one string swaps the
model — everything else stays the same, because the API is identical across all of them.

In [24]:
## Cell 11A — config (run once per session)
import os, getpass

API_KEY = os.environ.get("OPENROUTER_API_KEY") or getpass.getpass("OpenRouter API key: ")
MODEL = "nvidia/nemotron-3-ultra-550b-a55b:free"
print("key loaded:", API_KEY[:8] + "…", "| model:", MODEL)

OpenRouter API key:  ········


key loaded: sk-or-v1… | model: nvidia/nemotron-3-ultra-550b-a55b:free


### 11B — connection test

Sends about five tokens and asks the model to reply "ok". It exists to answer one question:
is the problem the service, or my code?

Reading the response code:

| Code | Meaning
|---|---
| 200 | working
| 401 | key rejected
| 429 | too many requests
| 5xx | their server broke

**About 429.** Free models are shared. Many people call the same one, and when the pool is
saturated everyone gets refused — including you, on your first request of the day. It is not
a quota you personally used up. The error body says `limit_source`: if it reads
`upstream_provider_shared_pool`, the model is simply busy and there is nothing to fix.

Models with several hosting providers behind them survive this better than models with one,
which is why a lower-rated model sometimes works when a higher-rated one does not.

In [26]:
## Cell 11B — connectivity test: is it the key, or the rate limit?
import requests

r = requests.post(
    "https://openrouter.ai/api/v1/chat/completions",
    headers={"Authorization": f"Bearer {API_KEY}"},
    json={"model": MODEL, "messages": [{"role": "user", "content": "Reply with: ok"}]},
    timeout=30,
)
print(r.status_code)

200


In [23]:
## Cell 11B2 — probe several models, report which are live
import requests, time

CANDIDATES = [
    "z-ai/glm-5.2:free",
    "nvidia/nemotron-3-ultra-550b-a55b:free",
    "google/gemma-4-31b-it:free",
    "google/gemma-4-26b-a4b-it:free",
]

for m in CANDIDATES:
    try:
        r = requests.post(
            "https://openrouter.ai/api/v1/chat/completions",
            headers={"Authorization": f"Bearer {API_KEY}"},
            json={"model": m, "max_tokens": 5,
                  "messages": [{"role": "user", "content": "Reply with: ok"}]},
            timeout=20,
        )
        if r.status_code == 200:
            print(f"OK        {m}   (provider: {r.json().get('provider')})")
        else:
            note = r.json().get("error", {}).get("metadata", {}).get("limit_source", "")
            print(f"{r.status_code:<9} {m}   {note}")
    except Exception as e:
        print(f"ERROR     {m}   {type(e).__name__}")
    time.sleep(1)          # don't trip the per-minute limit while probing

429       z-ai/glm-5.2:free   upstream_provider_shared_pool
OK        nvidia/nemotron-3-ultra-550b-a55b:free   (provider: Nvidia)
429       google/gemma-4-31b-it:free   upstream_provider_shared_pool
OK        google/gemma-4-26b-a4b-it:free   (provider: Darkbloom)


### 11C — the RAG answer

Where retrieval and generation join. Three steps:

1. `search()` returns the top chunks for the question
2. those chunks are formatted into a numbered SOURCES block
3. the block, the question, and a set of rules go to the model

**The rules are the whole point.** The system prompt tells the model to cite every claim,
to say it does not know rather than guess, and to copy figures exactly rather than
recalculate them. Without those instructions the model answers from general knowledge about
dentistry and accounting — fluently, confidently, and with no way to tell which parts came
from your document.

**`temperature=0`** makes the output deterministic: the same question returns the same
answer. Necessary if you ever want to test whether a change improved anything, since
otherwise you cannot tell a real improvement from random variation.

**Retry on 429 only.** Rate limiting is temporary and worth waiting out. A rejected key is
not — retrying that five times just wastes a minute before reporting the wrong problem.

The function returns the answer *and* the chunk ids it was given, so you can check whether
the model's citations match what it actually received. That check is what the next stage is
built on.

In [27]:
## Cell 11C — generation
import time

SYSTEM = """You answer questions about a dental practice valuation using ONLY the numbered \
sources provided. Follow these rules exactly:

1. Every factual claim must come from a source. Cite it inline as [sde-XX].
2. If the sources do not contain the answer, say so plainly. Do not fill the gap from \
general knowledge about valuation or accounting.
3. Quote figures exactly as written. Do not round, recalculate, or infer new numbers.
4. Be concise.
5. If the sources do not answer the question, reply only: "Not covered by the available
   sources." Do not describe what the sources do contain."""


def answer(question, k=3, retries=3):
    hits = search(question, k)
    context = "\n\n".join(
        f"[{c['id']}] {c['heading']}\n{c['text'].split(chr(10)*2, 1)[1]}"
        for _, _, c in hits
    )

    for attempt in range(retries):
        r = requests.post(
            "https://openrouter.ai/api/v1/chat/completions",
            headers={"Authorization": f"Bearer {API_KEY}"},
            json={"model": MODEL, "temperature": 0,
                  "messages": [{"role": "system", "content": SYSTEM},
                               {"role": "user",
                                "content": f"SOURCES:\n\n{context}\n\nQUESTION: {question}"}]},
            timeout=60,
        )
        if r.status_code == 200:
            return r.json()["choices"][0]["message"]["content"], \
                   [(round(s, 3), c["id"]) for s, _, c in hits]

        if r.status_code != 429:              # only 429 is worth retrying
            raise RuntimeError(f"{r.status_code}: {r.text[:300]}")

        wait = 15 * (attempt + 1)
        print(f"rate-limited, waiting {wait}s")
        time.sleep(wait)

    raise RuntimeError("still rate-limited after retries")


text, retrieved = answer("Why is the automobile add-back only 25 percent?")
print("RETRIEVED:", retrieved, "\n")
print(text)

RETRIEVED: [(0.546, 'sde-06'), (0.281, 'sde-10'), (0.278, 'sde-05')] 

The automobile add-back uses a flat 25% personal-use fraction that is an unsourced assumption — it has no cited authority and is applied as a convention rather than a derived figure [sde-06].


## 12. Guardrails — checking the answer against the sources

The model was *told* to cite everything and invent nothing. Cell 12 checks whether it
actually did.

This matters because a wrong answer looks exactly like a right one. The model writes
fluently either way, so "it sounded correct" is not verification.

**Three checks, all mechanical — no second model, no API call, instant:**

| Check | Catches |
|---|---|
| `fabricated_citations` | cites `[sde-99]`, or a chunk it was never given |
| `unsupported_numbers` | any figure not present in the retrieved text |
| `uncited` | answered with no citation and no "I don't know" |

The number check is the important one. If the model writes *"$8,432 of personal use"* and
that figure appears nowhere in the sources, it invented it — and invented financial figures
are the failure that actually costs you. Comparing the numbers in the answer against the
numbers in the sources catches it without judgment or guesswork.

Numbers are normalized before comparing, so `$9,114.00`, `9114`, and `9,114.0` all count as
the same value. Citation markers are stripped first — otherwise `[sde-06]` reads as the
number 6 and every answer fails.

**Two things this does not catch.**

It cannot tell whether the *prose* is accurate. An answer can cite the right chunk, use no
numbers, and still describe the rule backwards. Catching that needs a second model judging
the first — worth adding later, but it costs another API call and the judge can be wrong too.

And it flags legitimate arithmetic. If the model computes *"25% of $15,000 = $3,750"*, that
result is not in the sources, so it gets flagged. Correct behaviour here, since the system
prompt tells the model not to calculate — but it is a false positive, not a bug.

In [40]:
## Cell 12 — deterministic guardrails: verify the answer against what was retrieved

import re

CITE = re.compile(r'\[(sde-\d+)\]')
NUM  = re.compile(r'\$?\d[\d,]*(?:\.\d+)?%?')
ABSTAIN = re.compile(r"do(?: not|n't) (?:contain|have)|not (?:in|covered|available|present)"
                     r"|cannot answer|no information", re.I)

def _numbers(text):
    """Extract numbers, normalized so $9,114.00 / 9114 / 9,114.0 all compare equal."""
    out = set()
    for tok in NUM.findall(text):
        try:
            out.add(float(tok.replace('$', '').replace(',', '').rstrip('%')))
        except ValueError:
            pass
    return out


def check(answer, retrieved_ids, chunks_by_id):
    body   = CITE.sub(' ', answer)
    source = "\n".join(chunks_by_id[i]["text"] for i in retrieved_ids)

    cited     = set(CITE.findall(answer))
    abstained = bool(ABSTAIN.search(body))
    bad_cites = cited - set(retrieved_ids)
    unsupported = _numbers(body) - _numbers(source)

   # Numbers echoed from the question aren't disclosures — the user already had them.
    leaked = sorted(_numbers(body) - _numbers(question)) if abstained else []

    return {
        "cited":                sorted(cited),
        "fabricated_citations": sorted(bad_cites),
        "unsupported_numbers":  sorted(unsupported),
        "leaked_in_refusal":    leaked,
        "uncited":              not cited and not abstained,
        "abstained":            abstained,
        "pass": not bad_cites and not unsupported and not leaked
                and (bool(cited) or abstained),
    }


by_id = {c["id"]: c for c in chunks}

def ask(question, k=3, show_ids=False):
    text, retrieved = answer(question, k)
    ids = [cid for _, cid in retrieved]
    report = check(text, ids, by_id)

    print(f"RETRIEVED : {retrieved}")
    print(f"GUARDRAIL : {'PASS' if report['pass'] else 'FAIL'}")
    for key in ("fabricated_citations", "leaked_in_refusal", "unsupported_numbers", "uncited"):
        if report[key]:
            print(f"    {key}: {report[key]}")

    print()
    print(text if show_ids else render(text, by_id))
    return None          # nothing to echo

def render(answer_text, by_id):
    """Replace [sde-06] with a human-readable section reference."""
    def sub(m):
        c = by_id.get(m.group(1))
        return f"[{c['heading'].split('>')[-1].strip()}]" if c else m.group(0)
    return CITE.sub(sub, answer_text)   

ask("Did Harvey Seybold see more patients in 2023 than all other dentists?")

RETRIEVED : [(0.321, 'sde-14'), (0.295, 'sde-01'), (0.209, 'sde-07')]
GUARDRAIL : FAIL
    leaked_in_refusal: [100.0, 2023.0]

The provided sources do not contain any information about Harvey Seybold's patient count or visits in 2023, nor do they provide comparative visit data for other dentists. The only visit data mentioned is for "IC" records (100 visits) and "Hyg Hyg" records, but these are not identified as Harvey Seybold [8. Open items].
